src/Processors/Transforms/FilterTransform.cpp

```diff
@@ -175,7 +176,7 @@ void FilterTransform::transform(Chunk & chunk)
     /// If all the rows pass through the filter.
     if (num_filtered_rows == num_rows_before_filtration)
     {
-        if (!remove_filter_column)
+        if (!remove_filter_column && !filter_is_input_column)
```

Comment:
Actually it doesn't fix the problem in general case. E.g. in case when an expression that returns `UInt8` is used as filter:

```sql
SELECT number, number % 3 FROM numbers(10) WHERE number % 3

┌─number─┬─modulo(number, 3)─┐
│      1 │                 1 │
│      2 │                 1 │
│      4 │                 1 │
│      5 │                 1 │
│      7 │                 1 │
│      8 │                 1 │
└────────┴───────────────────┘
```

And it's hard to fix it without removing this optimization at all.

Reply:
Yes. Then I think we need to evaluate the trade-off here, whether fixing the bug or keeping the optimization is better in general case. How do you think?